### ISSUE TITLE
Re-sync contacts from master device

### ISSUE THREAD

**Issue by ra-rock** [AsamK/signal-cli#223]
In tickets [#12](https://github.com/AsamK/signal-cli/issues/12) and [#111](https://github.com/AsamK/signal-cli/issues/221), people expressed interest in sending contacts from the signal-cli instance to Signal Desktop or other devices.

However, I'm interested in syncing new contacts added to the master device, in this case a mobile phone,  after the original pairing to the signal-cli instance. If this feature is not available, I would be happy to try implementing on my own with some advice on where to start.

Thanks

**Comment by AsamK** [comment]
Normally the mobile app should send a contacts update sync message after it has refreshed its contacts list (in the Android app in the `DirectoryHelper::refreshDirectory()` method).

If that doesn't work, you can trigger a contacts update from signal-cli by calling the `requestSyncContacts` method.

**Comment by ra-rock** [comment]
Yeah that actually makes way more sense. On it.

**Comment by ra-rock** [comment]
After some more careful testing I do see the contacts are automatically syncing. It's good to know about requestSyncContacts, too. Now I'm wondering if in my branch I'd like to have the JSON somehow note there was a contacts sync, even if I don't actually output the entirety of the sync message. Or maybe I should for the sake of being thorough? 

This is what an incoming contacts sync message looks like at the moment:
```
{
  "envelope": {
    "source": "+[NUMBER]",
    "sourceDevice": 1,
    "relay": null,
    "timestamp": 1569760482422,
    "isReceipt": false,
    "dataMessage": null,
    "syncMessage": {
      "sentMessage": null,
      "destination": null,
      "blockedNumbers": null,
      "readMessages": null
    },
    "callMessage": null,
    "receiptMessage": null
  }
}
```

**Comment by AsamK** [comment]
To output the complete contact sync message is probably too much, as it can get quite large and signal-cli already stores the contacts in its config file.
Maybe some boolean `isContactSync` or rather a `type` field, that's set to CONTACT_SYNC. That would then also be useful for group sync and the request sync messages.

**Comment by AsamK** [comment]
If the enum can be nicely serialized to a json string than I'd prefer an enum. Otherwise a string is fine as well, as those classes are just used for serialization.

**Comment by ra-rock** [comment]
I want to double check that the sync type is mutually exclusive. So something like the following would work:

```
enum JsonSyncMessageType {
  CONTACTS_SYNC,
  GROUPS_SYNC,
  REQUEST_SYNC
}

class JsonSyncMessage {

    JsonSyncDataMessage sentMessage;
    List<String> blockedNumbers;
    List<ReadMessage> readMessages;
    JsonSyncMessageType type;

    JsonSyncMessage(SignalServiceSyncMessage syncMessage) {
        if (syncMessage.getSent().isPresent()) {
            this.sentMessage = new JsonSyncDataMessage(syncMessage.getSent().get());
        }
        if (syncMessage.getBlockedList().isPresent()) {
            this.blockedNumbers = syncMessage.getBlockedList().get().getNumbers();
        }
        if (syncMessage.getRead().isPresent()) {
            this.readMessages = syncMessage.getRead().get();
        }
        
        if (syncMessage.getContacts().isPresent()) {
            this.type = CONTACT_SYNC;
        } else if (syncMessage.getGroups().isPresent()) {
            this.type = GROUPS_SYNC;
        } else if (syncMessage.getRequest().isPresent()) {
            this.type = REQUEST_SYNC;
        }
    }
}
```

**Comment by AsamK** [comment]
yes, as it's currently implemented in signal-service-java, a sync message only has one type of content. So they're mutually exclusive.
